sphinx/testing/_internal/markers.py

```diff
@@ -0,0 +1,340 @@
+            err = '%r has duplicated values' % 'buildername'
+            pytest.fail(format_mark_failure('sphinx', err))
+        env['buildername'] = buildername
+    else:
+        buildername = env.setdefault('buildername', default_builder)
+
+    if not buildername:
+        pytest.fail(format_mark_failure('sphinx', 'missing builder name'))
+
+    check_mark_keywords('sphinx', SphinxMarkEnviron.__annotations__, env, node=node)
+    return env
+
+
+def _get_test_srcdir(
+    srcdir: str | None,
+    testroot: str | None,
+    shared_result: str | None,
+) -> str:
+    """Deduce the sources directory from the given arguments.
+
+    :param srcdir: An optional explicit source directory name.
+    :param testroot: An optional testroot ID to use.
+    :param shared_result: An optional shared result ID.
+    :return: The sources directory name *srcdir* (non-empty string).
+    """
+    check_mark_str_args('sphinx', srcdir=srcdir, testroot=testroot)
+    check_mark_str_args('test_params', shared_result=shared_result)
+
+    if srcdir is not None:
+        # the srcdir is explicitly given, so we use this name
+        # and we do not bother to make it unique (the user is
+        # responsible for that !)
+        return srcdir
+
+    if shared_result is not None:
+        # include the testroot id for visual purposes (unless it is
+        # not specified, which only occurs when there is no rootdir)
+        return f'{testroot}-{shared_result}' if testroot else shared_result
+
+    if testroot is None:
+        # neither an explicit nor the default testroot ID is given
+        pytest.fail('missing %r or %r parameter' % ('testroot', 'srcdir'))
+
+    return testroot
+
+
+def process_sphinx(
+    node: PytestNode,
+    session_temp_dir: str | os.PathLike[str],
+    testroot_finder: TestRootFinder,
+    default_builder: str,
+    default_isolation: NormalizableIsolation,
+    shared_result: str | None,
+) -> tuple[list[Any], SphinxInitKwargs]:
+    """Process the :func:`pytest.mark.sphinx` marker.
+
+    :param node: The pytest node to parse.
+    :param session_temp_dir: The session temporary directory.
+    :param testroot_finder: The testroot finder object.
+    :param default_builder: The application default builder name.
+    :param default_isolation: The isolation default policy.
+    :param shared_result: An optional shared result ID.
+    :return: The application positional and keyword arguments.
+    """
+    # 1. process pytest.mark.sphinx
+    env = _get_sphinx_environ(node, default_builder)
+    # 1.1a. deduce the isolation policy from freshenv if possible
+    freshenv: bool | None = env.pop('freshenv', None)
+    if freshenv is not None:
+        if 'isolate' in env:
+            err = '%r and %r are mutually exclusive' % ('freshenv', 'isolate')
+            pytest.fail(format_mark_failure('sphinx', err))
+
+        isolation = env['isolate'] = Isolation.always if freshenv else default_isolation
```

Comment:
Should `isolate=True` also imply `freshenv=True`?  (and if not, why not?)

Reply:
Yes and no. 

Since 'freshenv=True' means "you wipe out the environment" but it's as if you were fully isolated (that's what you want when you do some `-E`, you want to be clean from whatever cached thing), having `isolate=True` is kind of equivalent to having `freshenv=True`. 

The reason why I did not put `freshenv=True` -> `isolate=True` is that once you've isolated your sources, you assume you are... isolated. You wouldn't use `-E` the very first time you build your doc right? so that's the same idea.